library/psa_crypto.c

```diff
@@ -2853,6 +2855,140 @@ psa_status_t psa_mac_verify_finish( psa_mac_operation_t *operation,
 /* Asymmetric cryptography */
 /****************************************************************/
 
+psa_status_t psa_sign_message( mbedtls_svc_key_id_t key,
+                               psa_algorithm_t alg,
+                               const uint8_t * input,
+                               size_t input_length,
+                               uint8_t * signature,
+                               size_t signature_size,
+                               size_t * signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_status_t unlock_status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_slot_t *slot;
+    size_t hash_length;
+    uint8_t hash[PSA_HASH_MAX_SIZE];
+
+    *signature_length = 0;
+
+    if( ! PSA_ALG_IS_SIGN_MESSAGE( alg ) )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    if ( ! PSA_ALG_IS_HASH( PSA_ALG_SIGN_GET_HASH( alg ) ) )
+        return( PSA_ERROR_INVALID_ARGUMENT );
+
+    /* Immediately reject a zero-length signature buffer. This guarantees
+     * that signature must be a valid pointer. (On the other hand, the hash
+     * buffer can in principle be empty since it doesn't actually have
+     * to be a hash.) */
+    if( signature_size == 0 )
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    status = psa_get_and_lock_key_slot_with_policy( key, &slot,
+                                                    PSA_KEY_USAGE_SIGN_MESSAGE,
```

Comment:
@gilles-peskine-arm Shouldn't keys registered with `PSA_KEY_USAGE_SIGN_HASH` also be able to be used with `sign_message`, since `SIGN_HASH` is more permissive than `SIGN_MESSAGE`?

Asking in the interest of compatibility, i.e. keys that have so far been created with SIGN_HASH permission cannot be upgraded to SIGN_MESSAGE, so the next best thing would be to allow them to be used for the all-in-one operation too.

Reply:
Indeed, and likewise with `VERIFY`. Though this isn't the best place to implement it, since [according to the specification](https://armmbed.github.io/mbed-crypto/html/api/keys/policy.html?highlight=psa_key_usage_sign_hash#c.PSA_KEY_USAGE_SIGN_HASH)

> if an application sets the flag PSA_KEY_USAGE_SIGN_HASH when creating a key, then the key always has the permissions conveyed by PSA_KEY_USAGE_SIGN_MESSAGE, and the flag PSA_KEY_USAGE_SIGN_MESSAGE will also be present when the application queries the usage flags of the key.

So the right place to implement this is when setting the attributes of a new key. Otherwise `psa_get_key_attributes` won't do the right thing. This is part of the scope of https://github.com/ARMmbed/mbedtls/issues/3267, but it's up to @gabor-mezei-arm whether to do it in this pull request or in a follow-up.

Upgrading existing keys in persistent storage was not in scope, so I've filed a [separate issue](https://github.com/ARMmbed/mbedtls/issues/4368) for it.

Reply:
It will done in a separate PR.